docs:  New terms and updates to glossary (#21982)
huggingface/transformers: docs/source/en/glossary.mdx

[[1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]
```

### autoencoding models 

see [masked language modeling](#masked-language-modeling)

### autoregressive models

> see [causal language modeling](#causal-language-modeling)

Reviewer: I would keep those as these are sometimes used (a quick search in the repo shows quite a few instances).

Color images are made up of some combination of values in three channels - red, green, and blue (RGB) - and grayscale images only have one channel. In 🤗 Transformers, the channel can be the first or last dimension of an image's tensor: [`n_channels`, `height`, `width`] or [`height`, `width`, `n_channels`].

### causal language modeling

A pretraining task where the model reads the texts in order and has to predict the next word. It's usually done by

> reading the whole sentence but using a mask inside the model to hide the future tokens at a certain timestep.

Reviewer: This definitely need to be kept, we use it all the time in the course and the examples. The class is `AutoModelForCausalLM` etc.

Vision-based Transformers models split an image into smaller patches which are linearly embedded, and then passed as a sequence to the model. You can find the `patch_size` - or resolution - of the model in it's configuration.

### inference

> Inference is the process of evaluating a model on new data after the training process is complete.

Author: That's a good idea, I'll add that in

The base models ([`BertModel`]) do not accept labels, as these are the base transformer models, simply outputting
features.

### large language models (LLM)

> A generic term which refers to transformer language models (e.g. GPT-3, BERT) that were trained on a large quantity of data.

Reviewer 2: These might be more relevant examples of LLMs than BERT

```suggestion
A generic term that refers to transformer language models (GPT-3, BLOOM, OPT) that were trained on a large quantity of data.
```

The base models ([`BertModel`]) do not accept labels, as these are the base transformer models, simply outputting
features.

### large language models (LLM)

> A generic term that refers to transformer language models (GPT-3, BLOOM, OPT) that were trained on a large quantity of data.

Reviewer: Also linked to their size (in number of parameters) I believe.

## N

### Natural language generation (NLG)

All tasks related to generating text (for instance talk with transformers, translation).

### Natural language processing (NLP)

A generic way to say "deal with texts".

### Natural language understanding (NLU)

All tasks related to understanding what is in a text (for instance classifying the
whole text, individual words).

## P

### pipeline

> A pipeline is an abstraction referring to a series of steps that are executed in a specific order to process and transform data to be used by a downstream machine learning model for training or inference. Some example stages found in a pipeline might be data preprocessing, feature extraction, normalization and model training.

Reviewer 2: Pipelines in the context of 🤗 Transformers don't involve training, so it'd probably be good to remove that to avoid any confusion.

```suggestion
A pipeline in 🤗 Transformers is an abstraction referring to a series of steps that are executed in a specific order to preprocess and transform data and return a prediction from a model. Some example stages found in a pipeline might be data preprocessing, feature extraction, and normalization.
```

Author: Oh yeah good catch!

> For more details see [Pipelines for inference](https://huggingface.co/docs/transformers/pipeline_tutorial)

Reviewer 2: ```suggestion
For more details, see [Pipelines for inference](https://huggingface.co/docs/transformers/pipeline_tutorial).
```

## N

### Natural language generation (NLG)

> All tasks related to generating text (for instance talk with transformers, translation).

Reviewer 2: Did you mean Write With Transformers?

```suggestion
All tasks related to generating text (for instance, [Write With Transformers](https://transformer.huggingface.co/), translation).
```

Author: I didn't write this entry but good catch!

Absolute positional embeddings are selected in the range `[0, config.max_position_embeddings - 1]`. Some models use
other types of positional embeddings, such as sinusoidal position embeddings or relative position embeddings.

### preprocessing

> The task of cleaning and organizing raw data into a format that can be consumed easily by machine learning models. To gain a better idea of what preprocessing is, we recommend checking out the [Preprocess](https://huggingface.co/docs/transformers/preprocessing) article, which goes over a large range of examples.

Reviewer 2: ```suggestion
The task of preparing raw data into a format that can be easily consumed by machine learning models. For example, text is typically preprocessed by tokenization. To gain a better idea of what preprocessing looks like for other input types, check out the [Preprocess](https://huggingface.co/docs/transformers/preprocessing) tutorial.
```

## R

### recurrent neural network (RNN)

A type of model that uses a loop over a layer to process texts.

### representation learning

> A subfield of machine learning which focuses on learning meaningful representations of raw data. Some examples of representation learning techniques include the use of word embeddings, autoencoders, and Generative Adversarial Networks (GANs).

Reviewer 2: ```suggestion
A subfield of machine learning which focuses on learning meaningful representations of raw data. Some examples of representation learning techniques include word embeddings, autoencoders, and Generative Adversarial Networks (GANs).
```

Each element of the input finds out which other elements of the input they should attend to.

### semi-supervised learning

> A broad category of machine learning training techniques which look to leverage a small amount of labeled data with a larger quantity of unlabeled data to improve the accuracy of a model, differentiating it from [supervised learning](#supervised-learning) and [unsupervised learning](#unsupervised-learning).

Reviewer 2: ```suggestion
A broad category of machine learning training techniques that leverages a small amount of labeled data with a larger quantity of unlabeled data to improve the accuracy of a model, unlike [supervised learning](#supervised-learning) and [unsupervised learning](#unsupervised-learning).
```

Each element of the input finds out which other elements of the input they should attend to.

### semi-supervised learning

A broad category of machine learning training techniques that leverages a small amount of labeled data with a larger quantity of unlabeled data to improve the accuracy of a model, unlike [supervised learning](#supervised-learning) and [unsupervised learning](#unsupervised-learning).

> An example of an semi-supervised learning approach is "self-training", in which a model is trained on labeled data, and then used to make predictions on the unlabeled data. The portion of the unlabeled data that the model predicts with the most confidence gets added to the labeled dataset and used to retrain the model.

Reviewer: ```suggestion
An example of a semi-supervised learning approach is "self-training", in which a model is trained on labeled data, and then used to make predictions on the unlabeled data. The portion of the unlabeled data that the model predicts with the most confidence gets added to the labeled dataset and used to retrain the model.
```

In [convolution](#convolution) or [pooling](#pooling), the stride refers to the distance the kernel is moved over a matrix. A stride of 1 means the kernel is moved one pixel over at a time, and a stride of 2 means the kernel is moved two pixels over at a time.

### supervised learning

> A form of model training which makes direct use of labeled data to correct and instruct model performance. Data is fed into the model being trained, and its predictions are then compared to the known labels. The model then has its weights updated based on how incorrect its predictions were, and the process is repeated to try and optimize model performance.

Reviewer 2: ```suggestion
A form of model training that directly uses labeled data to correct and instruct model performance. Data is fed into the model being trained, and its predictions are compared to the known labels. The model updates its weights based on how incorrect its predictions were, and the process is repeated to optimize model performance.
```

Some models, like [`XLNetModel`] use an additional token represented by a `2`.

### transfer learning

> A technique which involves taking a model pretrained for a specific task and using it for something completely different. Instead of training a model from scratch, we can leverage knowledge obtained from an existing model as a starting point, which can speed up the learning process and reduce the amount of training data needed.

Reviewer 2: ```suggestion
A technique that involves taking a pretrained model and adapting it to a dataset specific to your task. Instead of training a model from scratch, you can leverage knowledge obtained from an existing model as a starting point. This speeds up the learning process and reduces the amount of training data needed.
```
